blocksuite/affine/shared/src/commands/model-crud/replace-selected-text-with-blocks.ts

```diff
@@ -0,0 +1,337 @@
+    nextBlock.model.text.insert(textToMove, 0);
+    blockModel.text.delete(splitOffset, blockModel.text.length - splitOffset);
+  }
+  return nextBlock.model;
+};
+
+const getSelectedBlocks = (
+  doc: any,
+  textSelection: TextSelection
+): BlockModel[] | null => {
+  const selectedBlocks: BlockModel[] = [];
+  const fromBlock = doc.getBlock(textSelection.from.blockId)?.model;
+  if (!fromBlock) return null;
+  selectedBlocks.push(fromBlock);
+
+  // If the selection spans multiple blocks, add the blocks in between
+  if (textSelection.to) {
+    const toBlock = doc.getBlock(textSelection.to.blockId)?.model;
+    if (!toBlock) return null;
+
+    if (fromBlock.id !== toBlock.id) {
+      let currentBlock = fromBlock;
+      while (currentBlock.id !== toBlock.id) {
+        const nextBlock = doc.getNext(currentBlock);
+        if (!nextBlock) break;
+        selectedBlocks.push(nextBlock);
+        currentBlock = nextBlock;
+      }
+    }
+  }
+
+  return selectedBlocks;
+};
+
+const deleteSelectedText = (doc: any, textSelection: TextSelection) => {
+  const selectedBlocks = getSelectedBlocks(doc, textSelection);
+  if (!selectedBlocks || selectedBlocks.length === 0) return null;
+
+  const firstBlock = selectedBlocks[0];
+  const lastBlock = selectedBlocks[selectedBlocks.length - 1];
+  const startOffset = textSelection.from.index;
+
+  if (textSelection.to) {
+    // Delete text from startOffset to the end in the first block
+    if (firstBlock.text) {
+      firstBlock.text.delete(startOffset, firstBlock.text.length - startOffset);
+    }
+
+    // Delete text from the beginning to endOffset in the last block
+    if (lastBlock.text) {
+      lastBlock.text.delete(textSelection.to.index, textSelection.to.length);
+    }
+
+    // Merge first block and last block
+    if (firstBlock.text && lastBlock.text) {
+      firstBlock.text.insert(lastBlock.text.toString(), startOffset);
+    }
+
+    // Delete the blocks in between
+    selectedBlocks.slice(1).forEach(block => {
+      doc.deleteBlock(block);
+    });
+  } else {
+    // Single block selection case
+    if (firstBlock.text) {
+      firstBlock.text.delete(startOffset, textSelection.from.length);
+    }
+  }
+
+  return { startBlockModel: firstBlock, startOffset };
+};
+
+export const replaceSelectedTextWithBlocksCommand: Command<{
+  textSelection: TextSelection;
+  blocks: BlockModel[];
+}> = (ctx, next) => {
+  const { textSelection, blocks, std } = ctx;
+  const doc = std.host.doc;
+
+  // Delete selected text and get startOffset
+  const result = deleteSelectedText(doc, textSelection);
+  if (!result) return next();
+  const { startBlockModel, startOffset } = result;
+
+  const parent = doc.getParent(startBlockModel.id);
+  if (!parent) return next();
+
+  const pattern = getBlocksPattern(blocks);
+  const index = parent.children.findIndex(x => x.id === startBlockModel.id);
+
+  match(pattern)
+    .with({ first: 'paragraph', multiple: false }, () => {
```

Comment:
Do we just need to check whether it is a paragraph? What about list and callout blocks?

Reply:
Optimized:

* `canMergeBlocks`：Determines if blocks can be merged based on their types
* `isMergableBlock`: Check if a block is mergeable in general

> `canMergeBlocks` & `isMergableBlock` should be decoupled with flavor in the future